documentation/manual/src/doc/guides/json-codecs.md

```diff
@@ -2,25 +2,82 @@
 
-## Use codecs at the algebra level
+The following diagram summarizes which algebra to use in which case:
 
-The most general algebra for defining JSON entities uses separate
-types for request and response entities:
+~~~ mermaid
+graph TB
+Doc(Do I want to publish documentation?)-->|yes| JsonSchemaEntities
+Doc -->|no| Codec(Do I want to implement both a client and a server?)
+Codec -->|yes| JsonEntitiesFromCodec
+Codec -->|no| JsonEntities
 
-~~~ scala src=../../../../../algebras/algebra/src/main/scala/endpoints/algebra/JsonEntities.scala#request-response-types
+style JsonSchemaEntities fill:#fff
+style JsonEntitiesFromCodec fill:#fff
+style JsonEntities fill:#fff
 ~~~
 
-However, in case your endpoint definitions are interpreted in multiple ways these types
-should be aligned so that requests and responses are consistently encoded, decoded and
-documented by client, server and documentation interpreters. A consequence of this is that an
-implicit instance of such codecs has to be available for your data types at the
-definition-site of your endpoints.
+The first question to ask is “do I want to publish documentation of my
+endpoints?”. In order to document the JSON entities of your requests and
+responses you need a JSON schema for them, that’s why you have to use the
+`JsonSchemaEntities` algebra. See [below](#jsonschemaentities) for more details.
 
-The next sections show how to use the two families of “codec”-based `JsonEntities` algebra
-specializations: `JsonEntitiesFromCodec` and `JsonSchemaEntities`.
+In case you don’t need to document the schema of your JSON entities, the
+second question to ask is “do I implement both a client and a server?”.
+If this is the case, then you need to be able to both encode and decode
+each JSON entity (an entity encoded by the server will be decoded by the
+client, and _vice versa_). To ensure that encoders and decoders are consistent
+together both have to be provided at the definition site of your endpoints, that’s
+why you should use the `JsonEntitiesFromCodec` algebra. See
+[below](#jsonentitiesfromcodec) for more details.
+
+Last, if you answered “no” to both questions, then you don’t need to use an algebra,
+you can directly use the `JsonEntities` interpreter for your specific client or server.
+With this approach, you will have to provide either an encoder or a decoder for each
+JSON entity on case-by-case basis: if you implement a client for an endpoint whose
+response contains a JSON entity, you will only need a decoder.
+
+The next section reminds general information about the `JsonEntities` hierarchy,
+and the remaining sections provide more details on how to use `JsonSchemaEntities`
+and `JsonEntitiesFromCodec`.
+
+## The `JsonEntities` hierarchy
+
+The following diagram shows the relations between the three aforementioned algebras,
+`JsonEntities`, `JsonEntitiesFromCodec`, and `JsonSchemaEntities`, and their relations
+with the other algebras:
+
+![](json-entities.svg)
```

Comment:
That one looks like this:

![hierarchy](https://user-images.githubusercontent.com/332812/64848256-16fd5500-d611-11e9-822f-3d955710475a.png)